### PR TITLE
Split populating and demangling of the symbols

### LINKED ISSUE #30: Rust 2021 demangling support
Currently it fails to demangle most of the functions.

```
rizin -A rust-clap-example.release
[0x100005e84]> afl~sym.__ZN
0x1000580a8    1 76           sym.__ZN4core3ops8function5impls80__LT_impl_u20_core..ops..function..FnOnce_LT_A_GT__u20_for_u20__RF_mut_u20_F_GT_9call_once17hee
0x1000580f8    6 132          sym.__ZN4core3ptr104drop_in_place_LT_alloc..vec..Vec_LT__LP_usize_C_std..backtrace_rs..symbolize..gimli..Mapping_RP__GT__GT_17hf8
0x1000581b0    3 32           sym.__ZN4core3ptr111drop_in_place_LT_alloc..raw_vec..RawVec_LT__LP_usize_C_std..backtrace_rs..symbolize..gimli..Mapping_RP__GT__G
0x1000581e0    1 4            sym.__ZN4core3ptr117drop_in_place_LT_std..sys_common..backtrace.._print_fmt.._u7b__u7b_closure_u7d__u7d_.._u7b__u7b_closure_u7d
0x100058330    4 44           sym.__ZN4core3ptr125drop_in_place_LT_addr2line..FrameIter_LT_gimli..read..endian_slice..EndianSlice_LT_gimli..endianity..LittleEn
0x10005835c    3 24           sym.__ZN4core3ptr126drop_in_place_LT_core..cell..UnsafeCell_LT_core..option..Option_LT_std..backtrace_rs..symbolize..gimli..mmap.
0x100058374    3 36   -> 20   sym.__ZN4core3ptr130drop_in_place_LT_gimli..read..dwarf..Dwarf_LT_gimli..read..endian_slice..EndianSlice_LT_gimli..endianity..Lit
0x10005899c    5 112          sym.__ZN4core3ptr92drop_in_place_LT_std..io..Write..write_fmt..Adapter_LT_std..sys..unix..stdio..Stderr_GT__GT_17h0981bbcea8dbcc3
0x100058a34   11 192          sym.__ZN4core3ptr93drop_in_place_LT_core..option..Option_LT_std..backtrace_rs..symbolize..gimli..Mapping_GT__GT_17h2ea33d423b71ae
0x100058b28    3 20           sym.__ZN4core3ptr95drop_in_place_LT_alloc..vec..Vec_LT_std..backtrace_rs..symbolize..gimli..LibrarySegment_GT__GT_17h89973ee0e59d
0x10005cfac    3 20           sym.__ZN4core3ptr118drop_in_place_LT_alloc..vec..Vec_LT__LP_gimli..common..DebugInfoOffset_C_gimli..common..DebugArangesOffset_RP
0x10005d0b8    1 4            sym.__ZN4core3ptr122drop_in_place_LT__RF_alloc..boxed..Box_LT_dyn_u20_core..error..Error_u2b_core..marker..Send_u2b_core..marker.
0x10005d114    9 152          sym.__ZN4core3ptr137drop_in_place_LT_gimli..read..dwarf..Unit_LT_gimli..read..endian_slice..EndianSlice_LT_gimli..endianity..Litt
0x10005d1c4    6 128          sym.__ZN4core3ptr146drop_in_place_LT_alloc..vec..Vec_LT_addr2line..ResUnit_LT_gimli..read..endian_slice..EndianSlice_LT_gimli..en
0x10005d2b8    3 36   -> 20   sym.__ZN4core3ptr159drop_in_place_LT_alloc..sync..ArcInner_LT_gimli..read..dwarf..Dwarf_LT_gimli..read..endian_slice..EndianSlice
0x10005e118   26 404          sym.__ZN99__LT_alloc..collections..btree..map..BTreeMap_LT_K_C_V_C_A_GT__u20_as_u20_core..ops..drop..Drop_GT_4drop17h8813af00e3eb
0x1000633bc    1 12           sym.__ZN4core3ptr101drop_in_place_LT_alloc..sync..Arc_LT_std..sync..mutex..Mutex_LT_alloc..vec..Vec_LT_u8_GT__GT__GT__GT_17h635e5
0x1000633dc    1 4            sym.__ZN4core3ptr120drop_in_place_LT__LP__RF_std..ffi..os_str..OsString_C__RF_core..option..Option_LT_std..ffi..os_str..OsString
0x100063be8   16 332          sym.__ZN3std2io5impls74__LT_impl_u20_std..io..Write_u20_for_u20_alloc..vec..Vec_LT_u8_C_A_GT__GT_14write_vectored17hde45469881818
0x100063d34    1 8            sym.__ZN3std2io5impls74__LT_impl_u20_std..io..Write_u20_for_u20_alloc..vec..Vec_LT_u8_C_A_GT__GT_17is_write_vectored17h6f6398f294
```

Demangling rules and algorithms are defined here:
- https://doc.servo.org/rustc_demangle/legacy/fn.demangle.html
- https://doc.servo.org/src/rustc_demangle/legacy.rs.html#49-98

Attaching debug, release binaries, along with the sources (updated from https://github.com/mytechtip/rust-clap-example):
[rust-2021.zip](https://github.com/rizinorg/rz-libdemangle/files/10342128/rust-2021.zip)

### LINKED ISSUE #3404: Language should be inferred before demangling the symbols
Currently, the demangling happens in the

```c
RzList /*<RzBinSymbol *>*/ *symbols(RzBinFile *bf)
```

function, which has to be implemented for every executable format by its plugin. The interface can be found at

https://github.com/rizinorg/rizin/blob/f4e15d232a407a72ff8d393f22ff06daf4205c6c/librz/include/rz_bin.h#L531

It is invoked in `rz_bin_object_set_items` as following

https://github.com/rizinorg/rizin/blob/f4e15d232a407a72ff8d393f22ff06daf4205c6c/librz/bin/bobj.c#L455

The invocation of `symbols` happens before the invocation of `rz_bin_language_detect`

https://github.com/rizinorg/rizin/blob/f4e15d232a407a72ff8d393f22ff06daf4205c6c/librz/bin/bobj.c#L527

This implies that `symbols` tries to guess the language for every single symbol instead of using the language detected in `rz_bin_language_detect`. This language is then used to resolve the demangler and demangle the symbol. As expected, the guesses can be incorrect and hence many symbols don't even get demangled by the correct demangler (the one associated with the executable's language).

The solution to this is not just to put the `rz_bin_language_detect` before the `symbols` function is called. This is because `rz_bin_language_detect` uses the symbols in the binfile (`binfile->o->symbols`) to deduce the language of the executable. And the symbols list is empty until the `symbols` function gets executed, which populates the list.

One way would be to populate the symbols list, but not demangle it yet. Then detect the language. Once we know the language, we can now demangle the symbols, without guessing the language for every symbol. This would require us to implement two functions (on which populates, other which demangles) for all the executable plugins (not so good).

I would like to hear about any neater and less disruptive solutions.

### PR DESCRIPTION
**Pull request by DMaroo**
**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

See #3404 for details.

Currently, the output for the test highlighted in rizinorg/rz-libdemangle#30 looks as follows

```
$ rizin -A rust-clap-example.release
[0x100005e84]> afl~sym.__ZN
0x100023424    1 8            sym.__ZN4core3ptr1101drop_in_place_LT_core..iter..adapters..map..map_fold_LT__RF_clap..builder..os_str..OsStr_C_alloc..borrow..Co
0x1000417d0    1 8            sym.__ZN4core3ptr1296drop_in_place_LT_core..iter..adapters..copied..copy_fold_LT__RF_str_C__LP__RP__C_core..iter..adapters..map..
```

Regarding why these two show up, it is likely that the demangler thinks that these two are invalid symbols. Which, to be fair, is a bug (?) in the Rust demangler, but the language inference and demangling part works perfectly fine.

**Test plan**

Green CI.

**Closing issues**

Fixes rizinorg/rz-libdemangle#30.
Closes #3404.

### CODE REVIEW

**Comment by wargio** [comment]
- Please add a test with the same file referenced in the `rz-libdemangle` issue
- One ObjC test is broken in regard to demangling:

```
[XX] db/formats/mach0/fatmach0 fatmach0 arm64e rebasing and stripping
RZ_NOPLUGINS=1 /private/var/folders/bc/yvvw2g2s4jlg9r8x7tf61xv800009d/T/woodpecker-local-693079402/rizinorg/rizin/prefix/bin/rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -a arm -b 64 -Qc 'pxr 0x18 @ 0x100004080' bins/mach0/modernobjc-noarm64
-- stdout
--- expected
+++ actual
@@ -1,3 +1,3 @@
-0x100004080 0x0000000100009270   p....... @ section.10.__DATA_CONST.__objc_classlist 4295004784 18.__DATA.__objc_data section.18.__DATA.__objc_data,class ViewController R 0x100009298
-0x100004088 0x00000001000092e8   ........ 4295004904 18.__DATA.__objc_data class AppDelegate R 0x1000092c0
-0x100004090 0x0000000100009338   8....... 4295004984 18.__DATA.__objc_data class SceneDelegate R 0x100009310
+0x100004080 0x0000000100009270   p....... @ section.10.__DATA_CONST.__objc_classlist 4295004784 18.__DATA.__objc_data section.18.__DATA.__objc_data,_OBJC_CLASS_$_ViewController R 0x100009298
+0x100004088 0x00000001000092e8   ........ 4295004904 18.__DATA.__objc_data _OBJC_CLASS_$_AppDelegate R 0x1000092c0
+0x100004090 0x0000000100009338   8....... 4295004984 18.__DATA.__objc_data _OBJC_CLASS_$_SceneDelegate R 0x100009310
```

**Comment by XVilka** [comment]
Ping?

**Comment by XVilka** [comment]
Any updates?

**Comment by XVilka** [comment]
@wargio please also add a test for that Rust demangling from https://github.com/rizinorg/rz-libdemangle/issues/30

File: librz/core/cbin.c [hunk 1]
```diff
@@ -31,6 +31,25 @@
...
+	char *symbolname; ///< display symbol name
+	char *libname; ///< name of the lib this symbol is specific to, if any
+	char *nameflag; ///< flag name for symbol
+	char *demname; ///< demangled raw symbol name
+	char *demflag; ///< flag name for demangled symbol
+	char *classname; ///< classname
+	char *classflag; ///< flag for classname
+	char *methname; ///< methods [class]::[method]
+	char *methflag; ///< methods flag sym.[class].[method]
+} BinSymNames;
```
Comment (XVilka) [inline librz/core/cbin.c:51]: What if we need to get that information from the API? Like it's currently done in https://github.com/rizinorg/rizin/blob/dev/test/integration/test_bin_symbols.c
It should not be a private structure. Maybe put it in `<rz_bin.h>` instead?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cbin.c (hunk 1 above)
Comment (wargio) [reply librz/core/cbin.c:51]: tbh, that structure was used only for printing and nothing else.
also this structure is kinda meh.
we would need to parse any sort of language to generate this

Comment (XVilka) [reply librz/core/cbin.c:51]: For printing is fine. This is very common use case to parse the binary and get information about symbols, imports, exports using the API. This feature should not be removed. If you have any suggestions how to make a better structure - please do, if not, please keep it as is.

Comment (XVilka) [reply librz/core/cbin.c:51]: Sorry, I misunderstood, I used another structure, not this one. Ignore this comment.

File: librz/include/rz_bin.h [hunk 2]
```diff
@@ -281,8 +281,10 @@ typedef struct rz_bin_object_t {
 	RzBinRelocStorage *relocs;
 	RzBinStrDb *strings;
 	RzList /*<RzBinClass *>*/ *classes;
-	HtPP *classes_ht;
-	HtPP *methods_ht;
+	HtPP *name_to_class_object; ///< map[char *]RzBinClass
+	HtPP *glue_to_class_method; ///< map[char *]RzBinSymbol
+	HtPP *glue_to_class_field; ///< map[char *]RzBinClassField
+	HtUP *vaddr_to_class_method; ///< map[ut64]RzBinSymbol
```
Comment (ret2libc) [inline librz/include/rz_bin.h:287]: i think we use a sliughtly different syntax like `HtUP /*<offset, List *<LocListEntry>*/ *loc_table`

File: librz/util/thread_hash_table.c [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	};
+
+#define th_ht_free(name, v) rz_th_##name##_free(v)
+#define th_ht_free_decl(name) \
+	RZ_API void rz_th_##name##_free(th_ht_type(name) * ht) { \
+		if (!ht) { \
+			return; \
+		} \
+		name##_free(ht->table); \
+		rz_th_lock_free(ht->lock); \
```
Comment (ret2libc) [inline librz/util/thread_hash_table.c:22]: is it possible this happens when some other thread is doing something on the Ht?

File: test/db/formats/dex [hunk 4]
```diff
@@ -216,35 +216,35 @@ EXPECT=<<EOF
...
-1   ---------- WEAK FUNC  java.io   void PrintStream.println(String)
-2   ---------- WEAK FUNC  java.lang void Integer.<init>(int)
-3   ---------- WEAK FUNC  java.lang String Integer.toString()
-4   ---------- WEAK FUNC  java.lang void Object.<init>()
-5   ---------- WEAK FUNC  java.lang void StringBuilder.<init>()
-6   ---------- WEAK FUNC  java.lang StringBuilder StringBuilder.append(String)
-7   ---------- WEAK FUNC  java.lang String StringBuilder.toString()
+nth vaddr      bind type  lib                     name                                       
+---------------------------------------------------------------------------------------------
+0   0x00000000 WEAK FIELD java.lang.System        System.out
```
Comment (XVilka) [inline test/db/formats/dex:240]: Isn't this an unnecessary duplication of the name?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/util/thread_hash_table.c (hunk 3 above)
Comment (wargio) [reply librz/util/thread_hash_table.c:22]: well, true that is a bad idea, but use-after-free is something should be considered at code level not at structure level in the free function

File: test/db/formats/dex (hunk 4 above)
Comment (wargio) [reply test/db/formats/dex:240]: no, is a dup due how the fields are created for the table.

**Comment by XVilka** [comment]
Was done by @wargio. Thanks for the initial implementation!

**Pull request closed without merging**